src/Init/Lean/Elab/Quotation.lean

```diff
@@ -6,67 +6,85 @@ Authors: Sebastian Ullrich
 import Init.Lean.Syntax
 import Init.Lean.Elab.ResolveName
+import Init.Lean.Elab.Term
 import Init.Lean.Parser -- TODO: remove after removing old elaborator
 
 namespace Lean
-namespace Elab
 
--- TODO: replace with quotations where possible
-private def const (n : Name) : Syntax :=
-Syntax.node `Lean.Parser.Term.id #[Syntax.ident none n.toString.toSubstring n [(n, [])]]
+/-- Reflect a runtime datum back to surface syntax (best-effort). -/
+class HasQuote (α : Type) :=
+(quote : α → Syntax)
 
-private def app (fn arg : Syntax) : Syntax :=
-Syntax.node `Lean.Parser.Term.app #[fn, arg]
+export HasQuote (quote)
 
-private def appN (fn : Syntax) (args : Array Syntax) : Syntax :=
-args.foldl app fn
+instance Syntax.HasQuote : HasQuote Syntax := ⟨id⟩
+instance String.HasQuote : HasQuote String := ⟨mkStxStrLit⟩
+instance Nat.HasQuote : HasQuote Nat := ⟨fun n => mkStxNumLit $ toString n⟩
 
--- TODO: unclear if the following functions could also be useful to other code
 private def quoteName : Name → Syntax
-| Name.anonymous => const `Lean.Name.anonymous
-| Name.str n s _ => appN (const `Lean.mkNameStr) #[quoteName n, mkStxStrLit s]
-| Name.num n i _ => appN (const `Lean.mkNameNum) #[quoteName n, mkStxNumLit $ toString i]
+| Name.anonymous => Unhygienic.run `(_root_.Lean.Name.anonymous)
+| Name.str n s _ => Unhygienic.run `(_root_.Lean.mkNameStr %%(quoteName n) %%(Lean.HasQuote.quote s))
+| Name.num n i _ => Unhygienic.run `(_root_.Lean.mkNameNum %%(quoteName n) %%(Lean.HasQuote.quote i))
+
+instance Name.HasQuote : HasQuote Name := ⟨quoteName⟩
+
+private def appN (fn : Syntax) (args : Array Syntax) : Syntax :=
+args.foldl (fun fn arg => Unhygienic.run `(%%fn %%arg)) fn
+
+instance Prod.HasQuote {α β : Type} [HasQuote α] [HasQuote β] : HasQuote (α × β) :=
+⟨fun ⟨a, b⟩ => Unhygienic.run `(_root_.Prod.mk %%(Lean.HasQuote.quote a) %%(Lean.HasQuote.quote b))⟩
+
+private def quoteList {α : Type} [HasQuote α] : List α → Syntax
+| [] =>      Unhygienic.run `(_root_.List.nil)
+| (x::xs) => Unhygienic.run `(_root_.List.cons %%(Lean.HasQuote.quote x) %%(quoteList xs))
 
-private def quoteList : List Syntax → Syntax
-| [] => const `List.nil
-| (x::xs) => appN (const `List.cons) #[x, quoteList xs]
+instance List.HasQuote {α : Type} [HasQuote α] : HasQuote (List α) := ⟨quoteList⟩
 
-private def quoteArray : Array Syntax → Syntax
-| xs => app (const `List.toArray) $ quoteList xs.toList
+instance Array.HasQuote {α : Type} [HasQuote α] : HasQuote (Array α) :=
+⟨fun xs => let stx := quote xs.toList; Unhygienic.run `(_root_.List.toArray %%stx)⟩
 
-private partial def quote (env : Environment) (msc : Syntax) : Syntax → Syntax
-| Syntax.ident info rawVal val preresolved =>
+namespace Elab
+namespace Term
+
+private partial def quoteSyntax {m : Type → Type} [Monad m] [MonadQuotation m] (env : Environment) : Nat → Syntax → m Syntax
+| _, Syntax.ident info rawVal val preresolved => do
   -- TODO: pass scope information
   let ns := Name.anonymous;
   let openDecls := [];
```

Comment:
I am trying to understand the code. Shouldn't we have the following two methods in the `MonadQuotation` class:
- `getCurrNamespace : m Name`
- `getOpenDecls : m (List OpenDecl)`

Reply:
This code is now in the `TermElabM` monad. It was generic only for the old frontend, but I'm now using the same monad there as well.
In general, the global lookup is done at compile time of the quotation, i.e. by the elaborator, while `MonadQuotation` is used to retrieve the macro scope at run time of the quotation. So the former is always run in the elaboration monad, only the latter should be flexible.